components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/HSMBasedExternalCryptoProvider.java

```diff
@@ -0,0 +1,534 @@
+     * This implementation supports only RSA private keys at the moment.
+     *
+     * @param cryptoContext  The context information which is used to discover the applicable private key.
+     * @param privateKeyInfo The information which is needed to retrieve the private key.
+     *                       If this information is not sufficient, the {@link CryptoContext} will be used to
+     *                       get more information.
+     * @return The {@link java.security.PrivateKey} relates with the given context.
+     * @throws CryptoException If something unexpected happens during private key discovery.
+     */
+    @Override
+    public java.security.PrivateKey getPrivateKey(CryptoContext cryptoContext,
+                                                  PrivateKeyInfo privateKeyInfo) throws CryptoException {
+
+        failIfContextInformationIsMissing(cryptoContext);
+        PrivateKey privateKeyTemplate = new PrivateKey();
+        privateKeyTemplate.getLabel().setCharArrayValue(privateKeyInfo.getKeyAlias().toCharArray());
+        privateKeyTemplate.getObjectClass().setLongValue(PKCS11Constants.CKO_PRIVATE_KEY);
+        Session session = initiateSession(slotResolver.resolveSlot(cryptoContext), false);
+        PrivateKey retrievedKey;
+        try {
+            retrievedKey = (PrivateKey) retrieveKey(privateKeyTemplate, session);
+        } finally {
+            sessionHandler.closeSession(session);
+        }
+        if (!retrievedKey.getSensitive().getBooleanValue() && retrievedKey.getExtractable().getBooleanValue()) {
+            return PKCS11JCEObjectMapper.mapPrivateKeyPKCS11ToJCE(retrievedKey);
+        } else {
+            String errorMessage = String.format("Requested private key %s is not extractable.",
+                    privateKeyInfo.getKeyAlias());
+            throw new CryptoException(errorMessage);
+        }
+    }
+
+    /**
+     * Computes and returns the {@link HybridEncryptionOutput} based on provided {@link HybridEncryptionInput}
+     * Hybrid encryption is carried out using the underlying HSM device.
+     * Session object is created for symmetric key for decryption.
+     *
+     * @param hybridEncryptionInput Input data for hybrid encryption.*
+     * @param symmetricAlgorithm    The symmetric encryption/decryption algorithm.
+     * @param asymmetricAlgorithm   The asymmetric encryption/decryption algorithm.
+     * @param javaSecurityProvider  The Java Security API provider. This value is discarded in this component.
+     * @param cryptoContext         The context information which is used to discover
+     *                              the public key of the external entity.
+     * @return {@link HybridEncryptionOutput} cipher text with required parameters
+     * @throws CryptoException
+     */
+    @Override
+    public HybridEncryptionOutput hybridEncrypt(HybridEncryptionInput hybridEncryptionInput, String symmetricAlgorithm,
```

Comment:
Can we split this into smaller methods ? Also we need more debug logs to see what's going on here.

Reply:
Added a separate method for key encryption.